File: simpleLinearKalman.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <random> 
#include <fstream>

struct Gaussian{
    double mean;
    double variance;
};

Gaussian update(double x_prior, double prior_variance, double x_measured, double measurement_variance)
{
    Gaussian posterior;
    posterior.mean = (prior_variance * x_measured + measurement_variance * x_prior)/(prior_variance + measurement_variance);
    posterior.variance = (prior_variance * measurement_variance)/(prior_variance + measurement_variance);

    return posterior;
}

Gaussian predict(double x_prior,double prior_variance, double velocity,double dt, double model_variance)
{   
    Gaussian predicted;
    predicted.mean = x_prior + velocity*dt;
    predicted.variance = prior_variance + model_variance;

    return predicted;
}

int main()
{
    // SET THE RANDOM GENERATOR SEED
    // Use current time as seed for random generator 
    srand(time(0)); 

    // GLOBAL SIMULATION SETTINGS
    constexpr double SIMULATION_STEPS{500};
    constexpr double dt{0.5};

    // OBJECT MOTION MODEL SETTINGS
    constexpr double x0{0.};
    constexpr double v0{1.};
    constexpr double PROCESS_VARIANCE{0.01};

    // SIMULATION TIMESTAMP VECTOR
    // PLACEHOLDER, I dont think I need this

    // TRUE OBJECT POSITION
    constexpr double PROCESS_NOISE{sqrt(PROCESS_VARIANCE)};

    //std::cout << "Process noise estimated at: " << PROCESS_NOISE << std::endl;

    // CALCULATE TRUE POSITIONS BASED ON MOTION MODEL
    // Initialize the result vector
    std::vector<double> X_real(SIMULATION_STEPS);

    // Set the first position to x0
    X_real[0] = x0;
    // Assign all the other elements
    double v{v0};

    // Get random double in range of 1. to -1.
    // Courtesy of Alessandro Jacopson
    // STACKOVERFLOW Q. 2704521
    std::uniform_real_distribution<double> unif(-1.,1.);
    std::default_random_engine re;

    for (int i{1};i<=SIMULATION_STEPS;i++)
        {
            v = v + PROCESS_NOISE * unif(re) * dt;
            X_real[i] = X_real[i-1] + v * dt;

        }

    // MEASUREMENT MODEL
    constexpr double MEASUREMENT_VARIANCE{100.};
    constexpr double MEASUREMENT_NOISE{sqrt(MEASUREMENT_VARIANCE)};
    // VECTOR WITH MEASURED SAMPLES
    std::vector<double> X_measured(SIMULATION_STEPS);

    // GENERATE MEASUREMENT WITH NOISE
    for (int i{0};i<=SIMULATION_STEPS;i++)
    {
        X_measured[i] = X_real[i] + unif(re) * MEASUREMENT_NOISE;

    }

    // KALMAN FILTER IMPLEMENTATION
    // Initial state estimate
    double x_prior{1.};
    constexpr double V_MODEL{0.5};
    constexpr double MODEL_VARIANCE{3.};

    // Initialize the output vector
    std::vector<double> X_result(SIMULATION_STEPS);

    // Step 0 - Initial step as predicted value, update KF output
    Gaussian posterior;
    Gaussian prior;

    posterior = update(x_prior,MODEL_VARIANCE,X_measured[0],MEASUREMENT_VARIANCE);

    X_result[0] = posterior.mean;

    // SET THE PRIOR TO THE POSTERIOR FROM PREVIOUS STEP
    prior = posterior;

    for (int i{1};i<=SIMULATION_STEPS;i++)
    {
        prior = predict(prior.mean,prior.variance,V_MODEL,dt,MODEL_VARIANCE);
        posterior = update(prior.mean,prior.variance,X_measured[i],MEASUREMENT_VARIANCE);
        // Append the result
        X_result[i] = posterior.mean;
        // UPDATE NEW PRIOR
        prior = posterior;
    } 

    // CONSOLE FORMATTING
    std::cout<<" KALMAN FILTERING RESULT "<<std::endl;

    // Save the output for python plotting
    std::ofstream outputFile;
    outputFile.open ("kf_result.csv");

    for (int i{1};i<=SIMULATION_STEPS;i++)
        {
            std::cout << "Real position is: " << X_real[i] <<" | Measured: "<<X_measured[i]<<" | Predicted: "<<X_result[i]<<std::endl;
            outputFile << X_real[i] <<","<<X_measured[i]<<","<<X_result[i]<<"\n";
        }

    outputFile.close();

}

File: simpleLinearKalman.py
# Simple Linear Kalman Filter
# Python implementation -> For .cpp version training
#--------------------------------------------------------------
# V 1.0
# BM
# 2024/05/11
#--------------------------------------------------------------
# Imports
from numpy.random import randn
from math import sqrt
import matplotlib.pyplot as plt
from numpy import linspace

# Global simulation settings
simulation_steps = 500
dt = 0.5

# "True" object model data
x0 = 0.
v = 1.
process_variance = 0.01 # Variability in the speed of tracked object

# Simulation timestamp vector
Time = linspace(0,(simulation_steps-1) * dt,simulation_steps)

# Calculate process noise 
process_noise = sqrt(process_variance)

# Calculate true position based on the motion model
X_real = [x0]
for i in range(simulation_steps-1):
    v = v + process_noise * randn() * dt
    x0 = X_real[i] + v * dt
    X_real.append(x0)

# Measurement model
measurement_variance = 100.
measurement_noise = sqrt(measurement_variance)

X_measured = []

# Append measurement samples with noise
for i in range(simulation_steps):
    X_measured.append(X_real[i] + randn()*measurement_noise)

# KALMAN FILTER IMPLEMENTATION
# Initial state estimate
x0 = 1.
v_model = 0.6
model_variance = 3

# KF Update function
def update(x_prior,prior_variance,x_measured,measurement_variance):
    x_posterior = (prior_variance * x_measured + measurement_variance * x_prior)/(prior_variance + measurement_variance)
    posterior_variance = (prior_variance * measurement_variance)/(prior_variance + measurement_variance)
    return x_posterior, posterior_variance

# KF Predict function
def predict(x_prior,prior_variance,velocity,dt,model_variance):
    x_predicted = x_prior + velocity*dt
    prediction_variance = prior_variance + model_variance
    return x_predicted, prediction_variance

# Calculate the estimates
result = []
#Update step 0 - initialize the KF
x_predict, x_variance = update(x0,model_variance,X_measured[0],measurement_variance)
result.append(x_predict)

#Main loop
for i in range(1,simulation_steps):
    # Predict
    x_prior, prior_variance = predict(x_predict, x_variance,v_model,dt,model_variance)
    # Update
    x_predict, x_variance = update(x_prior,prior_variance,X_measured[i],measurement_variance)
    #Append result
    result.append(x_predict)

# Plots
plt.scatter(Time,X_real,color='r',label='Real')
plt.scatter(Time,X_measured,color='b',label='Measurement')
plt.scatter(Time,result,color='g',label='Kalman')

plt.xlabel("Time / s")
plt.ylabel("Position / m")

plt.legend()
plt.show()